omigami/permutation_test.py

```diff
@@ -0,0 +1,154 @@
+
+    Examples
+    --------
+    >>> fs = FeatureSelector(n_outer=6, metric="MISS", estimator="PLSC")
+    >>> pt = PermutationTest(fs, 10)
+    >>> pt.fit(X, y)
+    >>> pt.compute_p_values(model="min")
+
+    """
+
+    def __init__(self, feature_selector: FeatureSelector, n_permutations: int = 20):
+        self.n_permutations = n_permutations
+        self._fs = feature_selector
+        self.res = None
+        self.res_perm = None
+
+    def fit(
+        self,
+        X: NumpyArray,
+        y: NumpyArray,
+        groups: NumpyArray = None,
+        executor: Executor = None,
+    ):
+
+        y_idx = np.arange(y.size)
+
+        if not self._fs.is_fit:
+            self._fs.fit(X, y, groups=groups)
+        self.res = self._get_feats_and_scores(self._fs)
+
+        fs_perm = FeatureSelector(**self._fs.get_params())
+        self.res_perm = []
+        for _ in tqdm.tqdm(range(self.n_permutations)):
+            np.random.shuffle(y_idx)
+            y_perm = y[y_idx]
+            fs_perm.fit(X, y_perm, groups=groups, executor=executor)
+            self.res_perm.append(self._get_feats_and_scores(fs_perm))
+
+        return self
+
+    @staticmethod
+    def _get_feats_and_scores(feature_selector: FeatureSelector) -> tuple:
+        selected_features = feature_selector.get_selected_features()
+        score = feature_selector.get_validation_curves()["total"][0]
+        return selected_features, score
+
+    def compute_permutation_scores(self, model: str) -> Tuple[float, List[float]]:
+        """Compute the permutation tests scores for input model. `model` must be one of
+        `"min"`, `"mid"` or `"max"`. It returns a tuple. The first element is the score
+        of the original featur selection, the second element is a list of scores, one
+        for each permutation
+
+        Parameters
+        ----------
+        model : str
+            one of `"min"`, `"mid"` or `"max"` models
+
+        Returns
+        -------
+        tuple
+            the original feature selection score and the scores for every permutation
+
+        Raises
+        ------
+        RuntimeError
+            if `fit` is not called first
+        ValueError
+            if `model` is not one of the three values allowed
+        """
+        if not self.res:
+            raise RuntimeError("Call fit method first")
+        if model not in {"min", "mid", "max"}:
+            return ValueError("Input model must be one of 'min', 'mid' or 'max'")
+        x = self._compute_model_score(model, self.res)
+        x_perm = [self._compute_model_score(model, r) for r in self.res_perm]
+        return x, x_perm
+
+    def compute_p_values(self, model: str, ranks: bool = False) -> float:
+        """Compute the p-value relative to the original feature selection score for
+        the input `model` based on the results of the various permutations.
+        If the permutation scores are not normal, set `ranks=True`.
```

Comment:
>How can we know about it?
Brunius in his tutorial suggests a visual inspection with a histogram, but I like your idea of using `scipy.stats.normaltest` better

Reply:
I set a threshold of 0.05 for the normaltest p value.
![immagine](https://user-images.githubusercontent.com/66323188/104733414-f41c0d00-573e-11eb-951b-a59b28f4ba39.png)
This is the average of the p value over 100 repetitions in function of the size of the sample popuation of a uniform distribution. It would take on average around 50 samples to raise the non-normality flag. Seems reasonable to me